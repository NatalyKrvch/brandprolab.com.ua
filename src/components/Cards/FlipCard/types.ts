import type { ReactNode } from 'react';

export interface FlipCardProps {
  front: ReactNode;
  back: ReactNode;
}
