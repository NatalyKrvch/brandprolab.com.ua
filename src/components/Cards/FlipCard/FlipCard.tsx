'use client';

import clsx from 'clsx';

import {
  FLIP_CARD_BACK_TEST_ID,
  FLIP_CARD_FRONT_TEST_ID,
  FLIP_CARD_TEST_ID,
} from '@/lib/testIDs';

import { useFlipCard } from './hooks/useFlipCard';
import type { FlipCardProps } from './types';

const FlipCard = ({ front, back, className }: FlipCardProps) => {
  const { flipped, handleFlip, handleKeyDown } = useFlipCard();

  return (
    <div
      className={clsx(
        'group relative cursor-pointer overflow-hidden [perspective:10000px]',
        className,
      )}
      role="button"
      tabIndex={0}
      aria-pressed={flipped}
      aria-label="Показати деталі"
      onClick={handleFlip}
      onKeyDown={handleKeyDown}
      data-testid={FLIP_CARD_TEST_ID}
    >
      <div
        className={clsx(
          'relative h-full w-full transition-all duration-500 [transform-style:preserve-3d]',
          flipped && '[transform:rotateX(-180deg)]',
        )}
      >
        <div
          className="absolute inset-0 flex h-full w-full [backface-visibility:hidden]"
          data-testid={FLIP_CARD_FRONT_TEST_ID}
        >
          {front}
        </div>

        <div
          className="absolute inset-0 flex h-full w-full [backface-visibility:hidden] [transform:rotateX(180deg)]"
          data-testid={FLIP_CARD_BACK_TEST_ID}
        >
          {back}
        </div>
      </div>
    </div>
  );
};

export default FlipCard;
