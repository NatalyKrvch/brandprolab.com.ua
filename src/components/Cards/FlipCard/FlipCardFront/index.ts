export { default as FlipCardFront } from './FlipCardFront';
