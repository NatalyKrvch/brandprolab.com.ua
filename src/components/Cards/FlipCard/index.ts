export { default as FlipCard } from './FlipCard';
export * from './types';
