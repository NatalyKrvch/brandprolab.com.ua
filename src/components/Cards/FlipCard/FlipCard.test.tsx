import '@testing-library/jest-dom';

import { fireEvent, render, screen } from '@testing-library/react';

import {
  FLIP_CARD_BACK_TEST_ID,
  FLIP_CARD_FRONT_TEST_ID,
  FLIP_CARD_TEST_ID,
} from '@/lib/testIDs';

import FlipCard from './FlipCard';

describe('FlipCard component', () => {
  const defaultProps = {
    front: <div>Front Side</div>,
    back: <div>Back Side</div>,
  };

  it('renders the FlipCard component', () => {
    const { asFragment } = render(<FlipCard {...defaultProps} />);
    expect(screen.getByTestId(FLIP_CARD_TEST_ID)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the front side content', () => {
    render(<FlipCard {...defaultProps} />);
    const frontElement = screen.getByTestId(FLIP_CARD_FRONT_TEST_ID);

    expect(frontElement).toBeInTheDocument();
    expect(frontElement).toHaveTextContent('Front Side');
  });

  it('renders the back side content', () => {
    render(<FlipCard {...defaultProps} />);
    const backElement = screen.getByTestId(FLIP_CARD_BACK_TEST_ID);

    expect(backElement).toBeInTheDocument();
    expect(backElement).toHaveTextContent('Back Side');
  });

  it('toggles class on click (flips the card)', () => {
    render(<FlipCard {...defaultProps} />);
    const wrapper = screen.getByTestId(FLIP_CARD_TEST_ID);
    const inner = wrapper.firstChild as HTMLElement;

    expect(inner).not.toHaveClass('[transform:rotateX(-180deg)]');

    fireEvent.click(wrapper);
    expect(inner).toHaveClass('[transform:rotateX(-180deg)]');

    fireEvent.click(wrapper);
    expect(inner).not.toHaveClass('[transform:rotateX(-180deg)]');
  });
});
