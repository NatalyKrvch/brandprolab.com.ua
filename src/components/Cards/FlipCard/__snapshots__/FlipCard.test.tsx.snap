// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlipCard component renders the FlipCard component 1`] = `
<DocumentFragment>
  <div
    class="group h-full w-full cursor-pointer [perspective:1000px]"
    data-testid="flip-card-test-id"
  >
    <div
      class="relative h-full w-full transition-all duration-500 [transform-style:preserve-3d]"
    >
      <div
        class="absolute flex h-full w-full [backface-visibility:hidden]"
        data-testid="flip-card-front-test-id"
      >
        Front Side
      </div>
      <div
        class="absolute flex h-full w-full [backface-visibility:hidden] [transform:rotateY(180deg)]"
        data-testid="flip-card-back-test-id"
      >
        Back Side
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`FlipCard component renders the back side content 1`] = `
<DocumentFragment>
  <div
    class="group h-full w-full cursor-pointer [perspective:1000px]"
    data-testid="flip-card-test-id"
  >
    <div
      class="relative h-full w-full transition-all duration-500 [transform-style:preserve-3d]"
    >
      <div
        class="absolute flex h-full w-full [backface-visibility:hidden]"
        data-testid="flip-card-front-test-id"
      >
        Front Side
      </div>
      <div
        class="absolute flex h-full w-full [backface-visibility:hidden] [transform:rotateY(180deg)]"
        data-testid="flip-card-back-test-id"
      >
        Back Side
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`FlipCard component renders the front side content 1`] = `
<DocumentFragment>
  <div
    class="group h-full w-full cursor-pointer [perspective:1000px]"
    data-testid="flip-card-test-id"
  >
    <div
      class="relative h-full w-full transition-all duration-500 [transform-style:preserve-3d]"
    >
      <div
        class="absolute flex h-full w-full [backface-visibility:hidden]"
        data-testid="flip-card-front-test-id"
      >
        Front Side
      </div>
      <div
        class="absolute flex h-full w-full [backface-visibility:hidden] [transform:rotateY(180deg)]"
        data-testid="flip-card-back-test-id"
      >
        Back Side
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`FlipCard component toggles between front and back when clicked 1`] = `
<DocumentFragment>
  <div
    class="group h-full w-full cursor-pointer [perspective:1000px]"
    data-testid="flip-card-test-id"
  >
    <div
      class="relative h-full w-full transition-all duration-500 [transform-style:preserve-3d]"
    >
      <div
        class="absolute flex h-full w-full [backface-visibility:hidden]"
        data-testid="flip-card-front-test-id"
      >
        Front Side
      </div>
      <div
        class="absolute flex h-full w-full [backface-visibility:hidden] [transform:rotateY(180deg)]"
        data-testid="flip-card-back-test-id"
      >
        Back Side
      </div>
    </div>
  </div>
</DocumentFragment>
`;
